File: src/main/java/io/github/ajoz/workshop/fp/part_1/solutions/exercise_5/Exercise5.java
package io.github.ajoz.workshop.fp.part_1.solutions.exercise_5;

@FunctionalInterface
interface Function1<A, B> {
    B apply(A a);
}

@FunctionalInterface
interface Consumer1<A> {
    void accept(A a);
}

@FunctionalInterface
interface Supplier<A> {
    A get();
}

class Exercise5 {
    static <A, B> Consumer1<A> composeConsumer(final Function1<A, B> function, final Consumer1<B> consumer) {
        return (A a) -> consumer.accept(function.apply(a));
    }

    static <A, B> Supplier<B> composeSupplier(final Function1<A, B> function, final Supplier<A> supplier) {
        return () -> function.apply(supplier.get());
    }

    static <A, B, C> Function1<B, C> applyFirst(final Function1<A, Function1<B, C>> function,
                                                final Supplier<A> supplier) {
        return (B b) -> function.apply(supplier.get()).apply(b);
    }
}

File: src/main/kotlin/io/github/ajoz/workshop/fp/kotlin/part_1/solutions/exercise_5/Exercise5.kt
@file:Suppress("PackageName")

package io.github.ajoz.workshop.fp.kotlin.part_1.solutions.exercise_5

fun <A, B> composeConsumer(function: (A) -> B, consumer: (B) -> Unit): (A) -> Unit = { a: A ->
    consumer(function(a))
}

fun <A, B> composeSupplier(function: (A) -> B, supplier: () -> A): () -> B = {
    function(supplier())
}

fun <A, B, C> applyCurriedFirst(function: (A) -> (B) -> C, supplier: () -> A): (B) -> C = { b: B ->
    function(supplier())(b)
}

fun <A, B, C> applyCurriedSecond(function: (A) -> (B) -> C, supplier: () -> B): (A) -> C = { a: A ->
    function(a)(supplier())
}